{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Bio 302 Homework 4\n",
    "\n",
    "All the problems from this week's notebook are in this notebook; there are no problems from the book, although the book will be helpful in answering some of the problems.\n",
    "\n",
    "**Submission**\n",
    "\n",
    "Submit your solutions as a link to a Gist via a Slack DM."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 1\n",
    "\n",
    "Write a function named `score_candidates`, that\n",
    "identifies the best-scoring sequence alignment(s) from a\n",
    "file containing some sequence alignments. The function\n",
    "takes three arguments:\n",
    "\n",
    "- the filename of the alignments file\n",
    "- the filename of the scoring matrix file\n",
    "- an integer gap penalty, typically negative\n",
    "\n",
    "A scoring matrix file looks like this:\n",
    "\n",
    "```\n",
    "A,C,G,T\n",
    "10\n",
    "-5,10\n",
    "0,-5,10\n",
    "-5,0,-5,10\n",
    "```\n",
    "\n",
    "This is the exact contents of the provided example\n",
    "scoring file, `dna_scores_1.txt`.\n",
    "\n",
    "What this data file corresponds to is a scoring table like this:\n",
    "\n",
    "```\n",
    "+---+----+----+----+----+\n",
    "|   | A  | C  | G  | T  |\n",
    "+---+----+----+----+----+\n",
    "|   |    |    |    |    |\n",
    "| A | 10 | -5 |  0 | -5 |\n",
    "|   |    |    |    |    |\n",
    "| C | -5 | 10 | -5 |  0 |\n",
    "|   |    |    |    |    |\n",
    "| G |  0 | -5 | 10 | -5 |\n",
    "|   |    |    |    |    |\n",
    "| T | -5 |  0 | -5 | 10 |\n",
    "+---+----+----+----+----+\n",
    "```\n",
    "\n",
    "It gives the scores for all possible alignments of the\n",
    "allowed characters in the sequence. For example,\n",
    "aligning a 'C' with 'C' adds 10 to your score, but\n",
    "aligning a 'C' with an 'A' adds -5 to your score.\n",
    "\n",
    "The first line of the data file contains the allowed\n",
    "characters in the sequence, separated by commas. There\n",
    "should then follow as many lines as there are\n",
    "characters, with scores separated by commas; here,\n",
    "there are four subsequent lines. It is implied that\n",
    "each successive line corresponds to the character in\n",
    "the header, so the '10' on its own corresponds to the\n",
    "'A'; the '-5,10' corresponds to 'C', and so on. Only\n",
    "the lower-left side of the matrix is given, because\n",
    "the matrices are symmetric.\n",
    "\n",
    "The alignments file has lines like this, with two\n",
    "aligned sequences separated by a comma, and no extra\n",
    "spaces:\n",
    "\n",
    "```\n",
    "TAGCAT---,TAC--GCAA\n",
    "TAGCAT---,TAC-G-CAA\n",
    "TAGCAT---,TACG--CAA\n",
    "TAGCAT--,-TACGCAA\n",
    "TAGCAT--,T-ACGCAA\n",
    "TAGCAT--,TA-CGCAA\n",
    "TAGCAT--,TAC-GCAA\n",
    "TAGCAT--,TACG-CAA\n",
    "TAGCAT--,TACGC-AA\n",
    "TAGCAT-,TACGCAA\n",
    "```\n",
    "\n",
    "Your function should compute the scores for each\n",
    "candidate alignment in the alignments file specified on the\n",
    "command line. The score should be computed as\n",
    "specified by the gap penalty and the scoring matrix\n",
    "given on the command line. In particular, the score of\n",
    "an alignment is simply the sum of the individual\n",
    "character-alignment scores, including gaps.\n",
    "\n",
    "Your function should return a dictionary. The dictionary\n",
    "should have two keys, one called `score`, which gives the value\n",
    "of the best score alignment, and a key called `alignments`,\n",
    "which is a list of length at least one, containing the winning\n",
    "candidate alignments(s).\n",
    "\n",
    "Examples:\n",
    "\n",
    "```\n",
    "In [7]: score_candidates('test_alignment_01.txt', 'dna_scores_1.txt', -4)\n",
    "Out[7]: {'alignments': [('T-AG', 'TGA-'), ('TAG-', 'T-GA')], 'score': 12}\n",
    "\n",
    "In [8]: score_candidates('test_alignment_02.txt', 'dna_scores_1.txt', -4)\n",
    "Out[8]: {'alignments': [('TAG', 'TAG')], 'score': 30}\n",
    "```\n",
    "\n",
    "If your function is correctly implemented, it should produce the same output as above. Demonstrate that this is the case in your solution."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "def score_candidates(file_alignments, file_scoring_matrix, gap_penalty):  \n",
    "    file = open(file_scoring_matrix).read()\n",
    "    header = file[:str.find(file,'\\n')].split(',')\n",
    "    length = len(header)\n",
    "    file = file[str.find(file,'\\n')+1:]\n",
    "    arr = np.zeros((length,length))\n",
    "    indices = np.tril_indices(length)\n",
    "    arr[indices] = file.strip().replace(',',' ').split()\n",
    "    arr = arr + arr.T - np.diag(arr.diagonal())\n",
    "\n",
    "    alignments = open(file_alignments).read().split('\\n')\n",
    "    candidates = dict(score=-1000, alignments=[])\n",
    "    for alignment in alignments:\n",
    "        score = 0\n",
    "        sequences = alignment.strip().split(',')\n",
    "        for i, char in enumerate(sequences[0]):\n",
    "            if(char == '-' or sequences[1][i] == '-'):\n",
    "                score += gap_penalty\n",
    "            else:\n",
    "                x = header.index(char)\n",
    "                y = header.index(sequences[1][i])\n",
    "                score += arr[x][y]\n",
    "        if (score == candidates['score']) and (alignment not in candidates['alignments']):\n",
    "            candidates['alignments'].append(alignment)\n",
    "        elif score > candidates['score']:\n",
    "            candidates['score'] = score\n",
    "            candidates['alignments'] = [alignment]\n",
    "    return candidates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'alignments': ['T-AG,TGA-', 'TAG-,T-GA'], 'score': 12.0}"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "score_candidates('test_alignment_01.txt','dna_scores_1.txt',-4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'alignments': ['TAG,TAG'], 'score': 30.0}"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "score_candidates('test_alignment_02.txt','dna_scores_1.txt',-4)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 2\n",
    "\n",
    "Use your function from Problem 1 to find the best-scoring alignments\n",
    "in the provided file `alignments_TAGCAT_TACGCAA.txt`. This file\n",
    "contains all possible alignments of the sequences\n",
    "`TAGCAT` and `TACGCAA`, in the format described above. Use the scoring matrix in\n",
    "provided file `dna_scores_1.txt` for all parts of this\n",
    "problem.\n",
    "\n",
    "(a) What are the winning alignment(s) when the gap\n",
    "    penalty is -4?\n",
    "\n",
    "(b) Experiment with gap penalties both smaller and larger than -4. Explain how and why the winning alignments change in relation to the gap penalty."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'score': 44.0, 'alignments': ['TA-GC-AT,TACGCAA-', 'TA-GCA-T,TACGCAA-', 'TA-GCAT-,TACGCA-A']}\n",
      "{'score': 41.0, 'alignments': ['TA-GCAT,TACGCAA']}\n",
      "{'score': 37.0, 'alignments': ['TA-GCAT,TACGCAA']}\n"
     ]
    }
   ],
   "source": [
    "print(score_candidates('alignments_TAGCAT_TACGCAA.txt','dna_scores_1.txt',-2))\n",
    "print(score_candidates('alignments_TAGCAT_TACGCAA.txt','dna_scores_1.txt',-4))\n",
    "print(score_candidates('alignments_TAGCAT_TACGCAA.txt','dna_scores_1.txt',-8))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The closer the gap penalty is to zero, the better alignments with gaps will score. This means that there will be more winning alignments with many gaps in them. The reverse is true as the gap penalty gets larger; the number of winning aligments will be restricted and fewer gaps will be present."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 3\n",
    "\n",
    "Make a dynamic programming table for the sequences in\n",
    "Problem 2, i.e. `TAGCAT` and `TACGCAA`. The gap\n",
    "penalty should be -4, and the scoring matrix should be\n",
    "as in `dna_scores_1.txt`.\n",
    "\n",
    "You can produce the table by hand or on a computer. In\n",
    "either case, make sure you show the traceback arrows\n",
    "in your diagram. Check out the markdown source in this cell\n",
    "to see how to include the picture in your notebook. (You'll either need to place the image file\n",
    "in the same directory as your notebook, or give the full path to the file.)\n",
    "\n",
    "![Hambone is a Shar Pug!](hambone.jpg)\n",
    "\n",
    "Is/are the best alignment(s) you produce using the dynamic programming table the same as in Problem 2a? Explain."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "![asda!](dynamic_programming_table.jpg)\n",
    "\n",
    "TACGCAA\n",
    "TA-GCAT\n",
    "\n",
    "The best alignment produced by the dynamic programming table is the same as the solution in problem 2a, which is just as it should be. Dynamic programming is only a shortcut to the same method as was used in 2a, which means that it should produce the same outcome."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 4\n",
    "\n",
    "One of the big advantages of the python language is that there is a wide variety of libraries available to augment the core python functionality. You can just install and `import` these libraries, and immediately make use of them.\n",
    "\n",
    "You've seen one such important external library already, `matplotlib`, which is the de facto python library for plotting. Another is `jupyter`, which provides the notebook in which you're reading this.\n",
    "\n",
    "Another important library is `biopython`. It provides a bunch of bioinformatics-related functionality. One of the many things it does is to provide reference versions of the common protein alignment substitution tables. Here's how to get at them. (These commands should work for you, because biopython is included in Anaconda; if they don't, let me know.)\n",
    "\n",
    "```\n",
    "In [60]: from Bio.SubsMat import MatrixInfo\n",
    "\n",
    "In [62]: print(MatrixInfo.available_matrices)\n",
    "['benner6', 'benner22', 'benner74', 'blosum100', 'blosum30', 'blosum35', 'blosum40', 'blosum45', 'blosum50', 'blosum55', 'blosum60', 'blosum62', 'blosum65', 'blosum70', 'blosum75', 'blosum80', 'blosum85', 'blosum90', 'blosum95', 'feng', 'fitch', 'genetic', 'gonnet', 'grant', 'ident', 'johnson', 'levin', 'mclach', 'miyata', 'nwsgappep', 'pam120', 'pam180', 'pam250', 'pam30', 'pam300', 'pam60', 'pam90', 'rao', 'risler', 'structure']\n",
    "```\n",
    "\n",
    "You'll recognize a bunch of these, the BLOSUM family and the PAM family. If you've imported `MatrixInfo`, you can access any of these matrices just by typing the name above after `MatrixInfo`. E.g. `MatrixInfo.blosum62` gives you the BLOSUM62 substitution matrix.\n",
    "\n",
    "Use biopython to print out and inspect the PAM250 scoring matrix.\n",
    "\n",
    "(a) What do large positive values in the matrix mean?\n",
    "\n",
    "(b) What do large negative values in the matrix mean?\n",
    "\n",
    "(c) What do zero values in the matrix mean?\n",
    "\n",
    "(d) What are the largest and smallest scores in the\n",
    "matrix, respectively? Does that make sense to you?\n",
    "Explain."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from Bio.SubsMat import MatrixInfo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['benner6', 'benner22', 'benner74', 'blosum100', 'blosum30', 'blosum35', 'blosum40', 'blosum45', 'blosum50', 'blosum55', 'blosum60', 'blosum62', 'blosum65', 'blosum70', 'blosum75', 'blosum80', 'blosum85', 'blosum90', 'blosum95', 'feng', 'fitch', 'genetic', 'gonnet', 'grant', 'ident', 'johnson', 'levin', 'mclach', 'miyata', 'nwsgappep', 'pam120', 'pam180', 'pam250', 'pam30', 'pam300', 'pam60', 'pam90', 'rao', 'risler', 'structure']\n"
     ]
    }
   ],
   "source": [
    "print(MatrixInfo.available_matrices)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{('A', 'A'): 4,\n",
       " ('B', 'A'): -2,\n",
       " ('B', 'B'): 4,\n",
       " ('B', 'C'): -3,\n",
       " ('B', 'D'): 4,\n",
       " ('B', 'E'): 1,\n",
       " ('B', 'F'): -3,\n",
       " ('B', 'G'): -1,\n",
       " ('B', 'H'): 0,\n",
       " ('B', 'I'): -3,\n",
       " ('B', 'K'): 0,\n",
       " ('B', 'L'): -4,\n",
       " ('B', 'M'): -3,\n",
       " ('B', 'N'): 3,\n",
       " ('B', 'P'): -2,\n",
       " ('B', 'Q'): 0,\n",
       " ('B', 'R'): -1,\n",
       " ('B', 'S'): 0,\n",
       " ('B', 'T'): -1,\n",
       " ('B', 'V'): -3,\n",
       " ('B', 'W'): -4,\n",
       " ('B', 'Y'): -3,\n",
       " ('C', 'A'): 0,\n",
       " ('C', 'C'): 9,\n",
       " ('C', 'D'): -3,\n",
       " ('C', 'N'): -3,\n",
       " ('C', 'R'): -3,\n",
       " ('D', 'A'): -2,\n",
       " ('D', 'D'): 6,\n",
       " ('D', 'N'): 1,\n",
       " ('D', 'R'): -2,\n",
       " ('E', 'A'): -1,\n",
       " ('E', 'C'): -4,\n",
       " ('E', 'D'): 2,\n",
       " ('E', 'E'): 5,\n",
       " ('E', 'N'): 0,\n",
       " ('E', 'Q'): 2,\n",
       " ('E', 'R'): 0,\n",
       " ('F', 'A'): -2,\n",
       " ('F', 'C'): -2,\n",
       " ('F', 'D'): -3,\n",
       " ('F', 'E'): -3,\n",
       " ('F', 'F'): 6,\n",
       " ('F', 'G'): -3,\n",
       " ('F', 'H'): -1,\n",
       " ('F', 'I'): 0,\n",
       " ('F', 'K'): -3,\n",
       " ('F', 'L'): 0,\n",
       " ('F', 'M'): 0,\n",
       " ('F', 'N'): -3,\n",
       " ('F', 'Q'): -3,\n",
       " ('F', 'R'): -3,\n",
       " ('G', 'A'): 0,\n",
       " ('G', 'C'): -3,\n",
       " ('G', 'D'): -1,\n",
       " ('G', 'E'): -2,\n",
       " ('G', 'G'): 6,\n",
       " ('G', 'N'): 0,\n",
       " ('G', 'Q'): -2,\n",
       " ('G', 'R'): -2,\n",
       " ('H', 'A'): -2,\n",
       " ('H', 'C'): -3,\n",
       " ('H', 'D'): -1,\n",
       " ('H', 'E'): 0,\n",
       " ('H', 'G'): -2,\n",
       " ('H', 'H'): 8,\n",
       " ('H', 'N'): 1,\n",
       " ('H', 'Q'): 0,\n",
       " ('H', 'R'): 0,\n",
       " ('I', 'A'): -1,\n",
       " ('I', 'C'): -1,\n",
       " ('I', 'D'): -3,\n",
       " ('I', 'E'): -3,\n",
       " ('I', 'G'): -4,\n",
       " ('I', 'H'): -3,\n",
       " ('I', 'I'): 4,\n",
       " ('I', 'N'): -3,\n",
       " ('I', 'Q'): -3,\n",
       " ('I', 'R'): -3,\n",
       " ('K', 'A'): -1,\n",
       " ('K', 'C'): -3,\n",
       " ('K', 'D'): -1,\n",
       " ('K', 'E'): 1,\n",
       " ('K', 'G'): -2,\n",
       " ('K', 'H'): -1,\n",
       " ('K', 'I'): -3,\n",
       " ('K', 'K'): 5,\n",
       " ('K', 'L'): -2,\n",
       " ('K', 'N'): 0,\n",
       " ('K', 'Q'): 1,\n",
       " ('K', 'R'): 2,\n",
       " ('L', 'A'): -1,\n",
       " ('L', 'C'): -1,\n",
       " ('L', 'D'): -4,\n",
       " ('L', 'E'): -3,\n",
       " ('L', 'G'): -4,\n",
       " ('L', 'H'): -3,\n",
       " ('L', 'I'): 2,\n",
       " ('L', 'L'): 4,\n",
       " ('L', 'N'): -3,\n",
       " ('L', 'Q'): -2,\n",
       " ('L', 'R'): -2,\n",
       " ('M', 'A'): -1,\n",
       " ('M', 'C'): -1,\n",
       " ('M', 'D'): -3,\n",
       " ('M', 'E'): -2,\n",
       " ('M', 'G'): -3,\n",
       " ('M', 'H'): -2,\n",
       " ('M', 'I'): 1,\n",
       " ('M', 'K'): -1,\n",
       " ('M', 'L'): 2,\n",
       " ('M', 'M'): 5,\n",
       " ('M', 'N'): -2,\n",
       " ('M', 'Q'): 0,\n",
       " ('M', 'R'): -1,\n",
       " ('N', 'A'): -2,\n",
       " ('N', 'N'): 6,\n",
       " ('N', 'R'): 0,\n",
       " ('P', 'A'): -1,\n",
       " ('P', 'C'): -3,\n",
       " ('P', 'D'): -1,\n",
       " ('P', 'E'): -1,\n",
       " ('P', 'F'): -4,\n",
       " ('P', 'G'): -2,\n",
       " ('P', 'H'): -2,\n",
       " ('P', 'I'): -3,\n",
       " ('P', 'K'): -1,\n",
       " ('P', 'L'): -3,\n",
       " ('P', 'M'): -2,\n",
       " ('P', 'N'): -2,\n",
       " ('P', 'P'): 7,\n",
       " ('P', 'Q'): -1,\n",
       " ('P', 'R'): -2,\n",
       " ('Q', 'A'): -1,\n",
       " ('Q', 'C'): -3,\n",
       " ('Q', 'D'): 0,\n",
       " ('Q', 'N'): 0,\n",
       " ('Q', 'Q'): 5,\n",
       " ('Q', 'R'): 1,\n",
       " ('R', 'A'): -1,\n",
       " ('R', 'R'): 5,\n",
       " ('S', 'A'): 1,\n",
       " ('S', 'C'): -1,\n",
       " ('S', 'D'): 0,\n",
       " ('S', 'E'): 0,\n",
       " ('S', 'F'): -2,\n",
       " ('S', 'G'): 0,\n",
       " ('S', 'H'): -1,\n",
       " ('S', 'I'): -2,\n",
       " ('S', 'K'): 0,\n",
       " ('S', 'L'): -2,\n",
       " ('S', 'M'): -1,\n",
       " ('S', 'N'): 1,\n",
       " ('S', 'P'): -1,\n",
       " ('S', 'Q'): 0,\n",
       " ('S', 'R'): -1,\n",
       " ('S', 'S'): 4,\n",
       " ('T', 'A'): 0,\n",
       " ('T', 'C'): -1,\n",
       " ('T', 'D'): -1,\n",
       " ('T', 'E'): -1,\n",
       " ('T', 'F'): -2,\n",
       " ('T', 'G'): -2,\n",
       " ('T', 'H'): -2,\n",
       " ('T', 'I'): -1,\n",
       " ('T', 'K'): -1,\n",
       " ('T', 'L'): -1,\n",
       " ('T', 'M'): -1,\n",
       " ('T', 'N'): 0,\n",
       " ('T', 'P'): -1,\n",
       " ('T', 'Q'): -1,\n",
       " ('T', 'R'): -1,\n",
       " ('T', 'S'): 1,\n",
       " ('T', 'T'): 5,\n",
       " ('V', 'A'): 0,\n",
       " ('V', 'C'): -1,\n",
       " ('V', 'D'): -3,\n",
       " ('V', 'E'): -2,\n",
       " ('V', 'F'): -1,\n",
       " ('V', 'G'): -3,\n",
       " ('V', 'H'): -3,\n",
       " ('V', 'I'): 3,\n",
       " ('V', 'K'): -2,\n",
       " ('V', 'L'): 1,\n",
       " ('V', 'M'): 1,\n",
       " ('V', 'N'): -3,\n",
       " ('V', 'P'): -2,\n",
       " ('V', 'Q'): -2,\n",
       " ('V', 'R'): -3,\n",
       " ('V', 'S'): -2,\n",
       " ('V', 'T'): 0,\n",
       " ('V', 'V'): 4,\n",
       " ('V', 'W'): -3,\n",
       " ('V', 'Y'): -1,\n",
       " ('W', 'A'): -3,\n",
       " ('W', 'C'): -2,\n",
       " ('W', 'D'): -4,\n",
       " ('W', 'E'): -3,\n",
       " ('W', 'F'): 1,\n",
       " ('W', 'G'): -2,\n",
       " ('W', 'H'): -2,\n",
       " ('W', 'I'): -3,\n",
       " ('W', 'K'): -3,\n",
       " ('W', 'L'): -2,\n",
       " ('W', 'M'): -1,\n",
       " ('W', 'N'): -4,\n",
       " ('W', 'P'): -4,\n",
       " ('W', 'Q'): -2,\n",
       " ('W', 'R'): -3,\n",
       " ('W', 'S'): -3,\n",
       " ('W', 'T'): -2,\n",
       " ('W', 'W'): 11,\n",
       " ('X', 'A'): 0,\n",
       " ('X', 'B'): -1,\n",
       " ('X', 'C'): -2,\n",
       " ('X', 'D'): -1,\n",
       " ('X', 'E'): -1,\n",
       " ('X', 'F'): -1,\n",
       " ('X', 'G'): -1,\n",
       " ('X', 'H'): -1,\n",
       " ('X', 'I'): -1,\n",
       " ('X', 'K'): -1,\n",
       " ('X', 'L'): -1,\n",
       " ('X', 'M'): -1,\n",
       " ('X', 'N'): -1,\n",
       " ('X', 'P'): -2,\n",
       " ('X', 'Q'): -1,\n",
       " ('X', 'R'): -1,\n",
       " ('X', 'S'): 0,\n",
       " ('X', 'T'): 0,\n",
       " ('X', 'V'): -1,\n",
       " ('X', 'W'): -2,\n",
       " ('X', 'X'): -1,\n",
       " ('X', 'Y'): -1,\n",
       " ('X', 'Z'): -1,\n",
       " ('Y', 'A'): -2,\n",
       " ('Y', 'C'): -2,\n",
       " ('Y', 'D'): -3,\n",
       " ('Y', 'E'): -2,\n",
       " ('Y', 'F'): 3,\n",
       " ('Y', 'G'): -3,\n",
       " ('Y', 'H'): 2,\n",
       " ('Y', 'I'): -1,\n",
       " ('Y', 'K'): -2,\n",
       " ('Y', 'L'): -1,\n",
       " ('Y', 'M'): -1,\n",
       " ('Y', 'N'): -2,\n",
       " ('Y', 'P'): -3,\n",
       " ('Y', 'Q'): -1,\n",
       " ('Y', 'R'): -2,\n",
       " ('Y', 'S'): -2,\n",
       " ('Y', 'T'): -2,\n",
       " ('Y', 'W'): 2,\n",
       " ('Y', 'Y'): 7,\n",
       " ('Z', 'A'): -1,\n",
       " ('Z', 'B'): 1,\n",
       " ('Z', 'C'): -3,\n",
       " ('Z', 'D'): 1,\n",
       " ('Z', 'E'): 4,\n",
       " ('Z', 'F'): -3,\n",
       " ('Z', 'G'): -2,\n",
       " ('Z', 'H'): 0,\n",
       " ('Z', 'I'): -3,\n",
       " ('Z', 'K'): 1,\n",
       " ('Z', 'L'): -3,\n",
       " ('Z', 'M'): -1,\n",
       " ('Z', 'N'): 0,\n",
       " ('Z', 'P'): -1,\n",
       " ('Z', 'Q'): 3,\n",
       " ('Z', 'R'): 0,\n",
       " ('Z', 'S'): 0,\n",
       " ('Z', 'T'): -1,\n",
       " ('Z', 'V'): -2,\n",
       " ('Z', 'W'): -3,\n",
       " ('Z', 'Y'): -2,\n",
       " ('Z', 'Z'): 4}"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MatrixInfo.blosum62"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "('W', 'W') 17\n",
      "('W', 'C') -8\n"
     ]
    }
   ],
   "source": [
    "pam250 = MatrixInfo.pam250\n",
    "max_score, min_score = 0,0\n",
    "ma = mi = []\n",
    "for k, v in pam250.items():\n",
    "    if(v > max_score):\n",
    "        max_score = v\n",
    "        ma = k\n",
    "    if(v < min_score):\n",
    "        min_score = v\n",
    "        mi = k\n",
    "print(ma,max_score)\n",
    "print(mi,min_score)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Use biopython to print out and inspect the PAM250 scoring matrix.\n",
    "(a) What do large positive values in the matrix mean?\n",
    "(b) What do large negative values in the matrix mean?\n",
    "(c) What do zero values in the matrix mean?\n",
    "(d) What are the largest and smallest scores in the matrix, respectively? Does that make sense to you? Explain.\n",
    "\n",
    "a) Large positive values in the matrix mean that the replacement between the two amino acids is likely to occur.\n",
    "b) Large negative values in the matrix mean that the two amino acids are very different and not likely to replace each other.\n",
    "c) Zero values in the matrix mean that the expected and observed occurrences of replacement between the two amino acids were equal.\n",
    "d)  ('W', 'W') 17\n",
    "    ('W', 'C') -8\n",
    "    These max and min make sense because tryptophan is a very unique amino acid, which means that it is very unlikely for it to be replaced by anything other than itself."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 5\n",
    "\n",
    "Use biopython to compare the PAM30 scoring matrix at\n",
    " with the PAM250 scoring\n",
    "matrix.\n",
    "\n",
    "(a) Overall, how do the entries in the two matrices\n",
    "differ from one another? You'll want to compare the\n",
    "diagonal entries between the two matrices, and then\n",
    "compare the off-diagonal entries. Give specific examples of amino acid pairs if\n",
    "that helps.\n",
    "\n",
    "(b) Explain the differences you see in the matrices,\n",
    "in terms of what you know about PAM30 and PAM250. Give\n",
    "examples of amino acid pairs if that helps."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{('A', 'A'): 6,\n",
       " ('B', 'A'): -3,\n",
       " ('B', 'B'): 6,\n",
       " ('B', 'C'): -12,\n",
       " ('B', 'D'): 6,\n",
       " ('B', 'E'): 1,\n",
       " ('B', 'F'): -10,\n",
       " ('B', 'G'): -3,\n",
       " ('B', 'H'): -1,\n",
       " ('B', 'I'): -6,\n",
       " ('B', 'K'): -2,\n",
       " ('B', 'L'): -9,\n",
       " ('B', 'M'): -10,\n",
       " ('B', 'N'): 6,\n",
       " ('B', 'P'): -7,\n",
       " ('B', 'Q'): -3,\n",
       " ('B', 'R'): -7,\n",
       " ('B', 'S'): -1,\n",
       " ('B', 'T'): -3,\n",
       " ('B', 'V'): -8,\n",
       " ('B', 'W'): -10,\n",
       " ('B', 'Y'): -6,\n",
       " ('C', 'A'): -6,\n",
       " ('C', 'C'): 10,\n",
       " ('C', 'D'): -14,\n",
       " ('C', 'N'): -11,\n",
       " ('C', 'R'): -8,\n",
       " ('D', 'A'): -3,\n",
       " ('D', 'D'): 8,\n",
       " ('D', 'N'): 2,\n",
       " ('D', 'R'): -10,\n",
       " ('E', 'A'): -2,\n",
       " ('E', 'C'): -14,\n",
       " ('E', 'D'): 2,\n",
       " ('E', 'E'): 8,\n",
       " ('E', 'N'): -2,\n",
       " ('E', 'Q'): 1,\n",
       " ('E', 'R'): -9,\n",
       " ('F', 'A'): -8,\n",
       " ('F', 'C'): -13,\n",
       " ('F', 'D'): -15,\n",
       " ('F', 'E'): -14,\n",
       " ('F', 'F'): 9,\n",
       " ('F', 'G'): -9,\n",
       " ('F', 'H'): -6,\n",
       " ('F', 'I'): -2,\n",
       " ('F', 'K'): -14,\n",
       " ('F', 'L'): -3,\n",
       " ('F', 'M'): -4,\n",
       " ('F', 'N'): -9,\n",
       " ('F', 'Q'): -13,\n",
       " ('F', 'R'): -9,\n",
       " ('G', 'A'): -2,\n",
       " ('G', 'C'): -9,\n",
       " ('G', 'D'): -3,\n",
       " ('G', 'E'): -4,\n",
       " ('G', 'G'): 6,\n",
       " ('G', 'N'): -3,\n",
       " ('G', 'Q'): -7,\n",
       " ('G', 'R'): -9,\n",
       " ('H', 'A'): -7,\n",
       " ('H', 'C'): -7,\n",
       " ('H', 'D'): -4,\n",
       " ('H', 'E'): -5,\n",
       " ('H', 'G'): -9,\n",
       " ('H', 'H'): 9,\n",
       " ('H', 'N'): 0,\n",
       " ('H', 'Q'): 1,\n",
       " ('H', 'R'): -2,\n",
       " ('I', 'A'): -5,\n",
       " ('I', 'C'): -6,\n",
       " ('I', 'D'): -7,\n",
       " ('I', 'E'): -5,\n",
       " ('I', 'G'): -11,\n",
       " ('I', 'H'): -9,\n",
       " ('I', 'I'): 8,\n",
       " ('I', 'N'): -5,\n",
       " ('I', 'Q'): -8,\n",
       " ('I', 'R'): -5,\n",
       " ('K', 'A'): -7,\n",
       " ('K', 'C'): -14,\n",
       " ('K', 'D'): -4,\n",
       " ('K', 'E'): -4,\n",
       " ('K', 'G'): -7,\n",
       " ('K', 'H'): -6,\n",
       " ('K', 'I'): -6,\n",
       " ('K', 'K'): 7,\n",
       " ('K', 'L'): -8,\n",
       " ('K', 'N'): -1,\n",
       " ('K', 'Q'): -3,\n",
       " ('K', 'R'): 0,\n",
       " ('L', 'A'): -6,\n",
       " ('L', 'C'): -15,\n",
       " ('L', 'D'): -12,\n",
       " ('L', 'E'): -9,\n",
       " ('L', 'G'): -10,\n",
       " ('L', 'H'): -6,\n",
       " ('L', 'I'): -1,\n",
       " ('L', 'L'): 7,\n",
       " ('L', 'N'): -7,\n",
       " ('L', 'Q'): -5,\n",
       " ('L', 'R'): -8,\n",
       " ('M', 'A'): -5,\n",
       " ('M', 'C'): -13,\n",
       " ('M', 'D'): -11,\n",
       " ('M', 'E'): -7,\n",
       " ('M', 'G'): -8,\n",
       " ('M', 'H'): -10,\n",
       " ('M', 'I'): -1,\n",
       " ('M', 'K'): -2,\n",
       " ('M', 'L'): 1,\n",
       " ('M', 'M'): 11,\n",
       " ('M', 'N'): -9,\n",
       " ('M', 'Q'): -4,\n",
       " ('M', 'R'): -4,\n",
       " ('N', 'A'): -4,\n",
       " ('N', 'N'): 8,\n",
       " ('N', 'R'): -6,\n",
       " ('P', 'A'): -2,\n",
       " ('P', 'C'): -8,\n",
       " ('P', 'D'): -8,\n",
       " ('P', 'E'): -5,\n",
       " ('P', 'F'): -10,\n",
       " ('P', 'G'): -6,\n",
       " ('P', 'H'): -4,\n",
       " ('P', 'I'): -8,\n",
       " ('P', 'K'): -6,\n",
       " ('P', 'L'): -7,\n",
       " ('P', 'M'): -8,\n",
       " ('P', 'N'): -6,\n",
       " ('P', 'P'): 8,\n",
       " ('P', 'Q'): -3,\n",
       " ('P', 'R'): -4,\n",
       " ('Q', 'A'): -4,\n",
       " ('Q', 'C'): -14,\n",
       " ('Q', 'D'): -2,\n",
       " ('Q', 'N'): -3,\n",
       " ('Q', 'Q'): 8,\n",
       " ('Q', 'R'): -2,\n",
       " ('R', 'A'): -7,\n",
       " ('R', 'R'): 8,\n",
       " ('S', 'A'): 0,\n",
       " ('S', 'C'): -3,\n",
       " ('S', 'D'): -4,\n",
       " ('S', 'E'): -4,\n",
       " ('S', 'F'): -6,\n",
       " ('S', 'G'): -2,\n",
       " ('S', 'H'): -6,\n",
       " ('S', 'I'): -7,\n",
       " ('S', 'K'): -4,\n",
       " ('S', 'L'): -8,\n",
       " ('S', 'M'): -5,\n",
       " ('S', 'N'): 0,\n",
       " ('S', 'P'): -2,\n",
       " ('S', 'Q'): -5,\n",
       " ('S', 'R'): -3,\n",
       " ('S', 'S'): 6,\n",
       " ('T', 'A'): -1,\n",
       " ('T', 'C'): -8,\n",
       " ('T', 'D'): -5,\n",
       " ('T', 'E'): -6,\n",
       " ('T', 'F'): -9,\n",
       " ('T', 'G'): -6,\n",
       " ('T', 'H'): -7,\n",
       " ('T', 'I'): -2,\n",
       " ('T', 'K'): -3,\n",
       " ('T', 'L'): -7,\n",
       " ('T', 'M'): -4,\n",
       " ('T', 'N'): -2,\n",
       " ('T', 'P'): -4,\n",
       " ('T', 'Q'): -5,\n",
       " ('T', 'R'): -6,\n",
       " ('T', 'S'): 0,\n",
       " ('T', 'T'): 7,\n",
       " ('V', 'A'): -2,\n",
       " ('V', 'C'): -6,\n",
       " ('V', 'D'): -8,\n",
       " ('V', 'E'): -6,\n",
       " ('V', 'F'): -8,\n",
       " ('V', 'G'): -5,\n",
       " ('V', 'H'): -6,\n",
       " ('V', 'I'): 2,\n",
       " ('V', 'K'): -9,\n",
       " ('V', 'L'): -2,\n",
       " ('V', 'M'): -1,\n",
       " ('V', 'N'): -8,\n",
       " ('V', 'P'): -6,\n",
       " ('V', 'Q'): -7,\n",
       " ('V', 'R'): -8,\n",
       " ('V', 'S'): -6,\n",
       " ('V', 'T'): -3,\n",
       " ('V', 'V'): 7,\n",
       " ('V', 'W'): -15,\n",
       " ('V', 'Y'): -7,\n",
       " ('W', 'A'): -13,\n",
       " ('W', 'C'): -15,\n",
       " ('W', 'D'): -15,\n",
       " ('W', 'E'): -17,\n",
       " ('W', 'F'): -4,\n",
       " ('W', 'G'): -15,\n",
       " ('W', 'H'): -7,\n",
       " ('W', 'I'): -14,\n",
       " ('W', 'K'): -12,\n",
       " ('W', 'L'): -6,\n",
       " ('W', 'M'): -13,\n",
       " ('W', 'N'): -8,\n",
       " ('W', 'P'): -14,\n",
       " ('W', 'Q'): -13,\n",
       " ('W', 'R'): -2,\n",
       " ('W', 'S'): -5,\n",
       " ('W', 'T'): -13,\n",
       " ('W', 'W'): 13,\n",
       " ('X', 'A'): -3,\n",
       " ('X', 'B'): -5,\n",
       " ('X', 'C'): -9,\n",
       " ('X', 'D'): -5,\n",
       " ('X', 'E'): -5,\n",
       " ('X', 'F'): -8,\n",
       " ('X', 'G'): -5,\n",
       " ('X', 'H'): -5,\n",
       " ('X', 'I'): -5,\n",
       " ('X', 'K'): -5,\n",
       " ('X', 'L'): -6,\n",
       " ('X', 'M'): -5,\n",
       " ('X', 'N'): -3,\n",
       " ('X', 'P'): -5,\n",
       " ('X', 'Q'): -5,\n",
       " ('X', 'R'): -6,\n",
       " ('X', 'S'): -3,\n",
       " ('X', 'T'): -4,\n",
       " ('X', 'V'): -5,\n",
       " ('X', 'W'): -11,\n",
       " ('X', 'X'): -5,\n",
       " ('X', 'Y'): -7,\n",
       " ('X', 'Z'): -5,\n",
       " ('Y', 'A'): -8,\n",
       " ('Y', 'C'): -4,\n",
       " ('Y', 'D'): -11,\n",
       " ('Y', 'E'): -8,\n",
       " ('Y', 'F'): 2,\n",
       " ('Y', 'G'): -14,\n",
       " ('Y', 'H'): -3,\n",
       " ('Y', 'I'): -6,\n",
       " ('Y', 'K'): -9,\n",
       " ('Y', 'L'): -7,\n",
       " ('Y', 'M'): -11,\n",
       " ('Y', 'N'): -4,\n",
       " ('Y', 'P'): -13,\n",
       " ('Y', 'Q'): -12,\n",
       " ('Y', 'R'): -10,\n",
       " ('Y', 'S'): -7,\n",
       " ('Y', 'T'): -6,\n",
       " ('Y', 'W'): -5,\n",
       " ('Y', 'Y'): 10,\n",
       " ('Z', 'A'): -3,\n",
       " ('Z', 'B'): 0,\n",
       " ('Z', 'C'): -14,\n",
       " ('Z', 'D'): 1,\n",
       " ('Z', 'E'): 6,\n",
       " ('Z', 'F'): -13,\n",
       " ('Z', 'G'): -5,\n",
       " ('Z', 'H'): -1,\n",
       " ('Z', 'I'): -6,\n",
       " ('Z', 'K'): -4,\n",
       " ('Z', 'L'): -7,\n",
       " ('Z', 'M'): -5,\n",
       " ('Z', 'N'): -3,\n",
       " ('Z', 'P'): -4,\n",
       " ('Z', 'Q'): 6,\n",
       " ('Z', 'R'): -4,\n",
       " ('Z', 'S'): -5,\n",
       " ('Z', 'T'): -6,\n",
       " ('Z', 'V'): -6,\n",
       " ('Z', 'W'): -14,\n",
       " ('Z', 'Y'): -9,\n",
       " ('Z', 'Z'): 6}"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pam30 = MatrixInfo.pam30\n",
    "pam30"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "a) The two matrices differ in that PAM30 gives more weight to replacement, which results in greater positive scores and negative scores.\n",
    "b) PAM30 is targeted more at similar sequences such as human vs chimpanzee beta globin, which explains why it would treat replacements more harshly and reward nonreplacement more strongly."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
